src/client/scripts/esm/game/rendering/draganimation.js

```diff
@@ -0,0 +1,131 @@
+ */
+
+/**
+ * To Do:
+ * - Emphasise the hovered square like Lichess and Chess.com.
+ */
+
+const z = 0.01;
+/** When not in perspective the pieces size is independent of board scale. */
+const touchscreenScale = 2;
+const mouseScale = 1;
+/** When using a touchscreen, the piece is shifted upward by this amount to prevent it being covered by fingers. */
+const touchscreenOffset = 2;
+/**
+ * The minimum size of the dragged piece relative to the stationary pieces.
+ * When zoomed in, this prevents it becoming tiny relative to the others.
+ */
+const minimumScale = 0.75;
+
+/** The hight the piece is rendered above the board when in perspective mode. */
+const perspectiveHeight = 0.6;
+const shadowColor = [0.1, 0.1, 0.1, 0.5];
+
+let startCoords;
+let endCoords;
+let pieceType;
+
+let transparentModel;
+let pieceModel;
+
+let hidden = false;
+
+function renderTransparentSquare() {
+	if(!startCoords) return;
+	genTransparentModel();
+	transparentModel.render();
+}
+
+function renderPiece() {
+	if(hidden || !endCoords) return;
+	genPieceModel();
+	pieceModel.render();
+}
+
+function genTransparentModel() {
+	let color = [0,0,0,0];
+	let data = shapes.getTransformedDataQuad_Color3D_FromCoord(startCoords, z, color); //Hide orginal piece
+	transparentModel = buffermodel.createModel_Colored(new Float32Array(data), 3, "TRIANGLES");
+}
+
+function genPieceModel() {
+	const perspectiveEnabled = perspective.getEnabled();
+	const touchscreen = input.getUsingTouchscreen();
+	const boardScale = movement.getBoardScale();
+	const rotation = perspective.getIsViewingBlackPerspective() ? -1 : 1;
+	
+	const { texleft, texbottom, texright, textop } = bufferdata.getTexDataOfType(pieceType, rotation);
+	const { r, g, b, a } = options.getColorOfType(pieceType);
+	const height = perspectiveEnabled ? perspectiveHeight * boardScale : z;
+	
+	let width = perspectiveEnabled ? boardScale : touchscreen ? touchscreenScale : mouseScale;
+	const minimumWidth = boardScale * minimumScale;
+	if (width < minimumWidth) width = minimumWidth; 
+	const left = endCoords[0] - width / 2;
+	const bottom = endCoords[1] - width / 2 + (touchscreen ? touchscreenOffset : 0);
+	const right = endCoords[0] + width / 2;
+	const top = endCoords[1] + width / 2 + (touchscreen ? touchscreenOffset : 0);
+	let data = [];
+	if (perspectiveEnabled) data.push(...bufferdata.getDataQuad_ColorTexture3D(left, bottom, right, top, z, texleft, texbottom, texright, textop, ...shadowColor));
+	data.push(...bufferdata.getDataQuad_ColorTexture3D(left, bottom, right, top, height, texleft, texbottom, texright, textop, r, g, b, a));
+	pieceModel = buffermodel.createModel_ColorTextured(new Float32Array(data), 3, "TRIANGLES", spritesheet.getSpritesheet());
+}
+
+/**
+ * 
+ * @param {string} type - The type of piece being dragged
+ * @param {number} pieceCoords - the square the piece was on
+ * @param {number[]} coords - the world coordinates the piece has been dragged to
+ */
+function dragPiece(type, pieceCoords, coords) {
```

Comment:
It's strange how you use dragPiece() to both initiate and update the position of the dragged piece. Could the logic for updating be moved into a new function?